Feature/content sentinel watch

santiagonoguez opened the pull request:
## Description
Add functionality to send watch requests to content sentinel, and handle FILE-UPDATE and FILE-ERROR responses.

## Motivation and Context
Part of current epic

## How Has This Been Tested?
Unit tests added for all new functionality.
E2E tests passed: https://app.circleci.com/pipelines/github/Rise-Vision/common-template/780/workflows/f64232ee-5113-4358-b2de-6a3828f80964
Tested manually with viewer-stage-2 using the following shared schedule: https://widgets.risevision.com/viewer-stage-2/?type=sharedschedule&id=4db24e51-1d5e-4468-b479-f573c584bf8a

Then from the console I manually watched a file with a console handler:

```
RisePlayerConfiguration.ContentSentinel.watchSingleFile( "risemedialibrary-f114ad26-949d-44b4-87e9-8528afc76ce4/santiago/unicorn.jpg", message => console.warn(JSON.stringify(message)) )

VM44:1 {"filePath":"risemedialibrary-f114ad26-949d-44b4-87e9-8528afc76ce4/santiago/unicorn.jpg","fileUrl":"risemedialibrary-f114ad26-949d-44b4-87e9-8528afc76ce4/santiago/unicorn.jpg","status":"CURRENT"}
```

And I also tested a nonexistent file:

```
RisePlayerConfiguration.ContentSentinel.watchSingleFile( "risemedialibrary-f114ad26-949d-44b4-87e9-8528afc76ce4/santiago/pato.jpg", message => console.warn(JSON.stringify(message)) )

VM59:1 {"filePath":"risemedialibrary-f114ad26-949d-44b4-87e9-8528afc76ce4/santiago/pato.jpg","fileUrl":null,"status":"FILE-ERROR","errorMessage":"HTTPS watch failed for risemedialibrary-f114ad26-949d-44b4-87e9-8528afc76ce4/santiago/pato.jpg"}
```

 - [x] Common template has been tested on Player - content sentinel is not activated here, so the previous lines have not effect.
 - [x] Common template has been tested in Apps Template Editor - with the template used in the presentation above - https://apps-stage-9.risevision.com/templates/edit/961e55e9-6582-42fb-af64-a6a78570eefa/?cid=121ae986-1b89-40c8-b10b-a64c60fb9e03

## Release Plan:
To be released before Friday.
Release is low risk as the new functionality is still not used from any component or template.
Simple rollback is enough to undo.

- As the Submitter, upon requesting review of this pull request, I confirm that the [Release Checklist](https://help.risevision.com/hc/en-us/articles/360031119991) has been completed.
- As the Reviewer, upon approving the changes in this PR, I confirm I have reviewed and I agree that the [Release Checklist](https://help.risevision.com/hc/en-us/articles/360031119991) has been completed

#### Release Checklist Items Skipped?
No need to notify support.

olegrise commented on src/rise-viewer.js:
```diff
@@ -102,6 +102,8 @@ RisePlayerConfiguration.Viewer = (() => {
       //forward logging messages from embedded temaplte to Viewer
       send( topic, message );
+    } else if ( topic === "file-update" || topic === "file-error" ) {
+      RisePlayerConfiguration.ContentSentinel.handleFileResponse( message )
```
@santiagonoguez do you think the message should be forwarded to the embedded templates?

santiagonoguez replied:
@olegrise I'm not familiar with embedded templates, but if a template has embedded templates and receives this message, yes, it should be forwarded. I would include that change in a different PR though, as goes beyond the scope of this one. Thanks !

olegrise commented on src/rise-content-sentinel.js:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  function watchSingleFile( filePath, handler ) {
+    RisePlayerConfiguration.Viewer.send( "watch", {
```
@santiagoarriaga The `RisePlayerConfiguration.Viewer.send` sends message to the parent window. In case of the embedded template the message won't reach the destination i.e. Viewer.

santiagonoguez replied:
@olegrise yes, good point. But then, shouldn't I fix that in RisePlayerConfiguration.Viewer.send instead ? because that function if I understand correctly is meant to always reack viewer, or am I missing something ?

santiagonoguez replied:
mmmmh, if I understand correctly I just need to resend like here, right ?
https://github.com/Rise-Vision/common-template/blob/feature/content-sentinel-watch/src/rise-viewer.js#L102

I'll add that and test with an embedded template, thanks !

olegrise replied:
@santiagoarriaga yes, exactly. 

Re. "shouldn't I fix that in RisePlayerConfiguration.Viewer.send instead ?" - not sure.

olegrise commented on src/rise-content-sentinel.js:
```diff
@@ -0,0 +1,83 @@
+    handler({
+      filePath: state.filePath,
+      fileUrl: null,
+      status: state.status,
```
@santiagoarriaga would it better to set status directly i.e. `status: "FILE-ERROR"` instead of changing the value of the object that was passed as a parameter?

santiagonoguez replied:
@olegrise I can make the change, thanks. I just tried to keep as close as possible to what's already implemented in rise-local-storage - https://github.com/Rise-Vision/common-template/blob/feature/content-sentinel-watch/src/rise-local-storage.js#L38

but I'll change both

santiagonoguez replied:
But then again, we'll still need to update state.status as that keeps track of the status changes for this handler.

olegrise commented on src/rise-content-sentinel.js:
```diff
@@ -0,0 +1,83 @@
+      return;
+    }
+
+    state.status = "FILE-ERROR";
```
@santiagonoguez maybe not a good idea to change value of the `state.status`
 because in theory there might be some other method down the road that need to know the original value.

santiagonoguez replied:
@olegrise the state object is only used in this file to keep track of the status of the file for the component requesting the watch, so no problem here. BTW, this implementation is derived from rise-local-storage, and that has been working without issues for some time.

olegrise commented on src/rise-content-sentinel.js:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  function _handleFileError( message, state, handler ) {
+    if ( !message || !message.filePath || message.filePath !== state.filePath ) {
```
@santiagonoguez is it possible for the `state` to be `null` or `undefined`?

santiagonoguez replied:
@olegrise it's not possible, because it's assigned as a non-empty object for every handler, and never cleared:

https://github.com/Rise-Vision/common-template/blob/feature/content-sentinel-watch/src/rise-content-sentinel.js#L50

santiagonoguez commented:
@olegrise @stulees embedded template considerations are now incorporated, using Oleg's suggested approach. I tested it running using a nested template with the same schedule under the same viewer-stage-2 environment ( same test URL, as for some reason viewer-stage-0 seems to not be running the content sentinel ). E2E and tests still pass, and I updated unit tests for embedded scenario. Please review again, thanks !

olegrise commented:
@santiagoarriaga I just noticed something strange. If I make watch request from the console `RisePlayerConfiguration.ContentSentinel.watchSingleFile( "risemedialibrary-f114ad26-949d-44b4-87e9-8528afc76ce4/santiago/unicorn.jpg", message => console.warn(JSON.stringify(message)) )`, then 20 seconds later that page starts consuming 100% CPU. Have you noticed that in your testing? I can make a video if you want.

stulees commented:
@santiagoarriaga Any reason to not test again production Viewer instead of stage-2? I'm concerned you're validating common template changes with a staging Viewer environment.

stulees commented on src/rise-content-sentinel.js:
```diff
@@ -0,0 +1,83 @@
+
+    Object.assign( state, { status, errorMessage: null, errorDetail: null });
+
+    // TO-DO create appropriate file URL for rise cache access
```
@santiagoarriaga Can you clarify this comment, why are you mentioning Rise Cache?

santiagonoguez replied:
@stulees ooops, I meant this is the key for Cache API ( not rise cache ). I'll fix, thanks !

santiagonoguez commented on src/rise-local-storage.js:
```diff
@@ -35,7 +35,7 @@ RisePlayerConfiguration.LocalStorage = (() => {
     handler({
       filePath: state.filePath,
       fileUrl: null,
-      status: state.status,
```
@stulees yes, I'll revert, and I'll leave the code for both files the same for now

santiagonoguez commented:
@olegrise that's the behavior I mentioned in the standup doc. It happens also if you don't watch, but slower. I'm not sure if it has to do with these changes. I'll do additional tests.

santiagonoguez commented:
@stulees at first I tried to user production viewer, but I couldn't find content sentinel running there. I'll give it a second try, thanks !

stulees commented on src/rise-content-sentinel.js:
```diff
@@ -0,0 +1,83 @@
+    Object.assign( state, { status, errorMessage: null, errorDetail: null });
+
+    // TO-DO create appropriate file URL for cache API access ( or leave it as filePath )
+    const fileUrl = status === "CURRENT" ? message.filePath : null;
```
@santiagoarriaga The value will always be `filePath` value, so do you think we should just omit `fileUrl`? See the description in my PR - https://github.com/Rise-Vision/rise-common-component/pull/76

santiagonoguez replied:
@stulees I'm wondering why this status === "CURRENT" is being done here. I assume is here to avoid trying to render when status is "STALE" or "UNKNOWN" ( which can come as status of file updated ). So maybe we still need this field to check if file should be rendered ( a valid up-to-date file path exists ) or not. Thoughts ? we can discuss tomorrow morning.

santiagonoguez commented:
@stulees I tested in viewer production, using a local charles proxy to divert calls to my common template version. Good news is that it worked as expected:

![viewer](https://user-images.githubusercontent.com/33162690/113933902-c03f0980-97ba-11eb-9f31-33300183ca99.png)

But bad news is that rogue behavior also happens here, after a few seconds, CPU goes to the top and after some time I start seeing errors in console:

![Screenshot from 2021-04-07 15-50-10](https://user-images.githubusercontent.com/33162690/113934057-eb295d80-97ba-11eb-8562-db5b3a1ca4c2.png)

santiagonoguez commented:
@stulees @olegrise I think this is now ready to be merged; as you know the content sentinel issue is now fixed. Please review this final code, thanks !

stulees reviewed (approved):
@santiagoarriaga Looks good to me 👍
